### ISSUE TITLE
Probe2 tests

### ISSUE THREAD

**Issue by russell-taylor** [ReliaSolve/cctbx_project#147]
Construct unit and regression tests for Probe2, placing them in the test directory and calling them tst_probe2_* and adding them to the mmtbx/run_tests.py script.

https://github.com/cctbx/cctbx_project/blob/master/CONTRIBUTING.md

**Comment by russell-taylor** [comment]
(done) Test a file with multiple models run using the self approach to make sure we get all models run by default. Then test with a model_id selection to make sure we only get the right one.  (nope, incompatible crystal volume) Use 1de3 to test this.  Others to try: 1axp, 1rrr, 1rrd; 1rrr can be read by probe2.

There are two meanings of multiple models in PDB files: (1) ensemble run, (2) multiple copies of the main set of atoms.  Probe looks like it just does each run separately (ensemble); verify that probe2 matches probe.

1rrr has ten structures, and the Probe2 -count output made ten different summaries.
`mmtbx.probe2 source_selection=all approach=self count_dots=True output.separate_worse_clashes=True output.file_name=C:\tmp\1rrr_new.txt F:/data/Richardsons/1rrr.cif`

The kinemage output made check-boxes for dots on M0 through M9.  The original Probe made models M1 through M10.
`mmtbx.probe2 source_selection="all" approach="self" count_dots=False output.add_kinemage_keyword=True output.separate_worse_clashes=True output.file_name=C:\tmp\1rrr_new.kin F:/data/Richardsons/1rrr.cif`

(done) Rename models M0-M9 to M1-M10.

(done) The number atoms selected is much different between Probe (6360) and Probe2 (636) per model when all models are run (seems like 636 would be the correct number for a single model).  The potential dots and area are also much different (around 30x larger).  When a single model is selected, Probe's numbers are much more similar to Probe2; exact atom counts and close other counts; this also fixes the score/A^2 because it is scaled by the correct value.  Probe is broken in this case, so we'll go with Probe2's counts.

drop_non_selected_atoms is no longer used for anything; its impact got lost somewhere along the way.  In fact, it was replaced by keep_unselected_atoms but was not removed.

(done) Running Probe2 with a "model 1" selection gets very different results than the first model output in the run with all atoms.  The "model 1" has much smaller counts (with most missing) than the Probe single-model or the Probe2 multi-model counts.  The atom counts and areas match, but the dot counts differ and the contact surface area differs.  This is probably because keep_unselected_atoms was not set to False.

(done) Do we want to handle the case of single-model selection by putting only the atoms in that model into the atom list rather than all atoms in the entire file?  This will require parsing the selection to find which model is being asked for and (off by one?) indexing into the model file to get that one.  This will be needed to enable the keep_unselected_atoms flag to do something meaningful in the case of a compound selection with model and something else; otherwise, it will always have to be False to avoid shielding by atoms from other models.  We added a check to ensure that an atom and its neighbor are in the same model before adding a neighbor to the potential collision list; this matches what Probe does.

**Comment by russell-taylor** [comment]
**@todo:** Why are we getting different radii for the Carbonyls on GLY 12 and GLY 36 in 1bti?  The new code seems to have a larger radius than the original.

**Comment by russell-taylor** [comment]
(done) Regression test the two new files Jane sent against the original Probe, using all-atom self test.

**Comment by russell-taylor** [comment]
(done) Add tests for the methods inside probe2.py along with regression tests of running it against some small molecules (one with water, one RNA, one self test without water).  Consider dumping and reporting kinemages if there is a difference and pointing to them in the error message.

**Comment by russell-taylor** [comment]
(done) Running on 4fenH produces many fewer H bonds against N and O, though similar amounts around C.  This could be because I removed the dummy hydrogens from interaction calculations.

Adding back the Phantom hydrogen interactions got about half of the missing H bonds back.

There are more contacts of other kinds against N.

These counts were very close when run on 3/21/2022 version.

**Comment by russell-taylor** [comment]
**@todo:** There are 92 C bad overlaps in Probe but 46 bad and 46 worse in Probe2 when running on 4fenH.

**Comment by russell-taylor** [comment]
(done) There are a lot of very bad clashes between N3 HPA 101 B and H3 U 51 B in Probe2 that are marked as Hydrogen bonds in Probe. (Not happening in 3/21/2022 version.)

N3 is marked as an acceptor in both the original and new Probe atom dumps when it is run on Linux.  H3 is marked as a donor in both of them.

**Comment by russell-taylor** [comment]
(done) There seem to be some O + H? collisions in 4fenH when run with -self against atoms that do not reciprocate with overlaps.  Check these against the original Probe and see if that is happening there as well.  (This may have been fixed with the neighbor chain fix.)

The radii are different for some of the HET residues in 4fenH for Probe and Probe2.

**Comment by russell-taylor** [comment]
(done) The O in HOH is marked as an acceptor in the new Reduce but not in the old during the comparison, but also the radius was 0... not all of the O are this way, only a subset of them (maybe half).  The others have a radius of 1.40 that matches the new ones.  There are the same number of atoms in both files.  It looks like we need to see what is happening with them so that they can be handled differently.

...except that Probe treats them all as both donors and acceptors in setProperties() and it seems like Reduce may be doing something different.

We're getting the same radii for Oxygen atoms on the 3/21/2022 version, no differences between O atoms in the Probe and Probe2 atom dumps.

**Comment by russell-taylor** [comment]
(done) N6 in A 17 (and other A's) of 4fenH is marked as a donor in the old probe but not in the new.  The two Hydrogens are present in the file we're testing.  N1 is not marked as a donor.  Neither has the atom marked as an acceptor.  There is the same difference in donor status for O2'.  There is an opposite donor difference for H02' (marked as donor in new but not in old).

Asked Jane how to handle this.  Mark all Nitrogens that have no Hydrogens in HET groups as donors.

**Comment by russell-taylor** [comment]
(done) Verify that the donors shifting to the Hydrogens from the heavy atoms in CCTBX is not breaking the Hydrogen-bond code.

The donor status was being adjusted this same way in the original probe; dumping the atom info after hydrogen adjustment verified this.

**Comment by russell-taylor** [comment]
(done) Many Hydrogens have radius 1.22 in the original Probe but they have radius 1.00 in the new: ARG HH11 HH12 HH21 HH22.  Some are now marked as donors in new, which is expected.

Dumping the atoms after hydrogen updates in the original probe shows that the radii are now 1.05 vs. 1.00 and the donor status always matches.

**Comment by russell-taylor** [comment]
(done) ARG N and NE; ALA N are marked as donors in original Probe and not in new; they all have hydrogens.  Testing with reduced 1bti. 

This is to be expected... the attached Hydrogens are the donors rather than the Nitrogens.

**Comment by russell-taylor** [comment]
(done) Probe2 regression test fails on reduced 1xso with a bad lookup on an atom type.  Also 3v7e.

Working on 3/21/2022 version.

**Comment by russell-taylor** [comment]
(done) Why are we getting differences in donor status on N, O atoms in ARG, LEU, SEP, SER?: Found in some combination of 1bti, 1crn, 1pq7_fragment, 6t5h_fragment.  Could be because of the fragments!
```
A ARG 238 N - flags differ: old is isDonor, new is noDonor
A ARG 238 NE - flags differ: old is isDonor, new is noDonor
A ARG 238 NH1 - flags differ: old is isDonor, new is noDonor
A ARG 238 NH2 - flags differ: old is isDonor, new is noDonor
A LEU 241 N - flags differ: old is isDonor, new is noDonor
A LEU 242 N - flags differ: old is isDonor, new is noDonor
A SEP 240 N - flags differ: old is isDonor, new is noDonor
A SEP 240 O - flags differ: old is isDonor, new is noDonor
A SEP 240 O1P - flags differ: old is isDonor, new is noDonor
A SEP 240 O2P - flags differ: old is isDonor, new is noDonor
A SEP 240 O3P - flags differ: old is isDonor, new is noDonor
A SEP 240 OG - flags differ: old is isDonor, new is noDonor
A SER 239 N - flags differ: old is isDonor, new is noDonor
A SER 239 OG - flags differ: old is isDonor, new is noDonor
```

Probably because of the fragments.

**Comment by russell-taylor** [comment]
(done) After updating chem_data by doing a fresh Phenix build from today's source, I can no longer run on 4fenH... Sorry unknown bonding energy type for 14 atoms: HPA B atoms C2, C4, C5, C6, C8, N1, N3, N7, N9, O6 missing. HPA is present in both the old and new chem_data.  The HPA.cif files are identical, as are the data_HPA.cif files... Setting MMTBX_CCP4_MONOMER_LIB to the old one gets it running even with the new CLIBD_MON.

I am able to run Hydrogenate with the new files.  Reduce2 gets a different error when run with the new MMTBX_CCP4_MONOMER_LIB (NoneType for an atom_dict entry in getExtraAtomInfo):
```
md, ani = mon_lib_srv.get_comp_comp_id_and_atom_name_interpretation(
                residue_name=ag.resname, atom_names=ag.atoms().extract_name())
```
This function returns a valid md when using an older MMTBX_CCP4_MONOMER_LIB and None when using a newer one.

Asked the Phenix developer mailing list about this.

Removed this call (whose results were no longer being used) and it now ends with a Sorry.  Tried running with hydrogenate and it also fails to find info on HPA, CU and ZN.  Sent an email to Dorothee asking for help.

The file /c/tmp/cctbx_phenix_old/modules/chem_data/chemical_components/chemical_components_data.py no longer exists in the new version.  Probably Phenix is using a different way to find the list that CCTBX does not and CCTBX is still trying to use the old list, which is not there.  That file is a list of ligands and their descriptions, including HPA.  Told Nigel about this.

Nigel reports fixing this.

**Comment by russell-taylor** [comment]
(done) Fix tst_probe2.py so that it works.

**Comment by russell-taylor** [comment]
(done) Remove the Phantom Hydrogen radius adjustments from probe2.py once the regression testing is done.

In fact, this needed to be done to make the regression tests behave correctly.  The adjustment was changing the behavior of the two cases.

**Comment by russell-taylor** [comment]
Running with the ener_lib_molprobity.cif file in place of the ener_lib.cif file within Phenix produces results for both 1bti_reduced and 4fenH that are very close to each other in values.  There remain some differences between the atoms in the two cases that could explain the differences in results:

```
1bti_reduced.orig.out:
program: probe.2.21.211221, run Fri Mar  4 10:58:16 2022
command: ./probe/probe -quiet -kin -mc -self all -count -sepworse -DUMPATOMS outputs/1bti_reduced.orig.dump outputs/temp_file.tmp
selection: self
name: dots
density: 16.0 dots per A^2
probeRad: 0.250 A
VDWrad: (r * 1.000) + 0.000 A
score weights: gapWt=0.25, bumpWt=10, HBWt=4
        
subgroup: self dots
atoms selected: 1043
potential dots: 122896
potential area: 7681.0 A^2
  type                 #      %       score score/A^2 x 1000
  C wide_contact    4425   3.6%      43.9      5.71
  C close_contact    3498   2.8%     152.2     19.82
  C small_overlap    1407   1.1%     -53.1     -6.92
  C bad_overlap      35   0.0%      -4.8     -0.63
  C worse_overlap      21   0.0%      -3.7     -0.48
  C H-bond      50   0.0%       0.7      0.10
Car wide_contact     167   0.1%       1.7      0.22
Car close_contact      77   0.1%       3.2      0.42
Car small_overlap      18   0.0%      -0.7     -0.10
  N wide_contact     901   0.7%       9.4      1.22
  N close_contact    1016   0.8%      47.0      6.12
  N small_overlap     281   0.2%     -11.0     -1.43
  N bad_overlap      11   0.0%      -1.5     -0.20
  N worse_overlap      13   0.0%      -2.3     -0.30
  N H-bond     982   0.8%      23.9      3.11
  O wide_contact    1777   1.4%      17.9      2.34
  O close_contact    1895   1.5%      85.6     11.14
  O small_overlap     597   0.5%     -22.0     -2.87
  O bad_overlap       4   0.0%      -0.5     -0.07
  O worse_overlap       4   0.0%      -0.7     -0.09
  O H-bond    2576   2.1%      71.1      9.25
  S wide_contact     234   0.2%       2.2      0.29
  S close_contact     199   0.2%       8.7      1.13
  S small_overlap      73   0.1%      -2.6     -0.34
  S bad_overlap       6   0.0%      -0.9     -0.11
  S worse_overlap       1   0.0%      -0.2     -0.02

     tot contact:    14189  11.5%     371.8     48.40
     tot overlap:     2471   2.0%    -104.2    -13.56
     tot  H-bond:     3608   2.9%      95.7     12.46

       grand tot:    20268  16.5%     363.3     47.30

contact surface area: 1266.8 A^2
```

```
1bti_reduced.new.out:
program: probe2 v.0.9.1, run 2022-03-04 10:58:20
command: /home/taylorr/rlab/phenix/build/../modules/cctbx_project/mmtbx/command_line/probe2.py source_selection="all" record_added_hydrogens=False approach=self count_dots=True output.separate_worse_clashes=True output.file_name=outputs/1bti_reduced.new.out output.dump_file_name=outputs/1bti_reduced.new.dump outputs/temp_file.tmp
selection: self
name: dots
density: 16.0 dots per A^2
probeRad: 0.250 A
VDWrad: (r * 1.000) + 0.000 A
score weights: gapWt=0.25, bumpWt=10, HBWt=4
        
subgroup: self dots
atoms selected: 1048
potential dots: 122897
potential area: 7681.1 A^2
  type                 #      %       score score/A^2 x 1000
  C wide_contact      4598   3.7%      45.6      5.94
  C close_contact     3597   2.9%     156.4     20.36
  C small_overlap     1423   1.2%     -53.8     -7.01
  C bad_overlap         35   0.0%      -4.8     -0.63
  C worse_overlap       21   0.0%      -3.7     -0.48
  C H-bonds             57   0.0%       0.8      0.10
  N wide_contact       901   0.7%       9.4      1.22
  N close_contact     1018   0.8%      47.1      6.13
  N small_overlap      281   0.2%     -11.0     -1.43
  N bad_overlap         11   0.0%      -1.5     -0.20
  N worse_overlap       13   0.0%      -2.3     -0.30
  N H-bonds            981   0.8%      23.9      3.11
  O wide_contact      1774   1.4%      17.9      2.33
  O close_contact     1897   1.5%      85.6     11.15
  O small_overlap      596   0.5%     -22.0     -2.87
  O bad_overlap          4   0.0%      -0.5     -0.07
  O worse_overlap        4   0.0%      -0.7     -0.09
  O H-bonds           2576   2.1%      70.9      9.23
  S wide_contact       233   0.2%       2.2      0.29
  S close_contact      199   0.2%       8.7      1.13
  S small_overlap       73   0.1%      -2.6     -0.34
  S bad_overlap          6   0.0%      -0.9     -0.11
  S worse_overlap        1   0.0%      -0.2     -0.02
  S H-bonds              7   0.0%       0.1      0.01

     tot contact:    14217  11.6%     372.9     48.55
     tot overlap:     2468   2.0%    -104.1    -13.55
     tot  H-bond:     3621   2.9%      95.6     12.45

       grand tot:    20306  16.5%     364.4     47.44

contact surface area: 1269.1 A^2
```

```
cat 1bti_reduced.compare
A ASN 24 CG - radii differ: old is 1.65, new is 1.70
A ASN 43 CG - radii differ: old is 1.65, new is 1.70
A ASN 44 CG - radii differ: old is 1.65, new is 1.70
A ASP 3 CG A radii differ: old is 1.65, new is 1.70
A ASP 3 CG B radii differ: old is 1.65, new is 1.70
A ASP 50 CG - radii differ: old is 1.65, new is 1.70
A GLN 31 CD - radii differ: old is 1.65, new is 1.70
A GLU 49 CD - radii differ: old is 1.65, new is 1.70
A GLU 7 CD - radii differ: old is 1.65, new is 1.70
```

```
cat outputs/4fenH.orig.out 
program: probe.2.21.211221, run Fri Mar  4 10:58:21 2022
command: ./probe/probe -quiet -kin -mc -self all -count -sepworse -DUMPATOMS outputs/4fenH.orig.dump outputs/temp_file.tmp
selection: self
name: dots
density: 16.0 dots per A^2
probeRad: 0.250 A
VDWrad: (r * 1.000) + 0.000 A
score weights: gapWt=0.25, bumpWt=10, HBWt=4
        
subgroup: self dots
atoms selected: 4799
potential dots: 497484
potential area: 31092.8 A^2
  type                 #      %       score score/A^2 x 1000
  C wide_contact   14369   2.9%     147.3      4.74
  C close_contact   13496   2.7%     594.7     19.13
  C small_overlap    3757   0.8%    -115.4     -3.71
  C bad_overlap      46   0.0%      -6.4     -0.21
  C worse_overlap      46   0.0%      -9.1     -0.29
  C H-bond    1253   0.3%      27.5      0.88
Car wide_contact     825   0.2%       9.2      0.30
Car close_contact     969   0.2%      42.2      1.36
Car small_overlap     236   0.0%      -6.6     -0.21
Car bad_overlap       1   0.0%      -0.1     -0.00
  N wide_contact    4889   1.0%      46.6      1.50
  N close_contact    3224   0.6%     139.1      4.48
  N small_overlap     405   0.1%      -9.9     -0.32
  N bad_overlap       6   0.0%      -0.8     -0.03
  N worse_overlap      12   0.0%      -2.3     -0.07
  N H-bond    8469   1.7%     239.2      7.69
  O wide_contact   10772   2.2%     112.3      3.61
  O close_contact   11908   2.4%     532.2     17.12
  O small_overlap    3317   0.7%     -95.3     -3.07
  O bad_overlap      34   0.0%      -4.7     -0.15
  O worse_overlap      52   0.0%     -10.1     -0.32
  O H-bond   35363   7.1%    1053.8     33.89
  P wide_contact     392   0.1%       2.8      0.09
  P close_contact     137   0.0%       5.7      0.18
  P small_overlap      19   0.0%      -0.3     -0.01

     tot contact:    60981  12.3%    1632.1     52.49
     tot overlap:     7931   1.6%    -261.0     -8.39
     tot  H-bond:    45085   9.1%    1320.4     42.47

       grand tot:   113997  22.9%    2691.5     86.56

contact surface area: 7124.8 A^2
```

```
cat outputs/4fenH.new.out 
program: probe2 v.0.9.1, run 2022-03-04 10:58:35
command: /home/taylorr/rlab/phenix/build/../modules/cctbx_project/mmtbx/command_line/probe2.py source_selection="all" record_added_hydrogens=False approach=self count_dots=True output.separate_worse_clashes=True output.file_name=outputs/4fenH.new.out output.dump_file_name=outputs/4fenH.new.dump outputs/temp_file.tmp
selection: self
name: dots
density: 16.0 dots per A^2
probeRad: 0.250 A
VDWrad: (r * 1.000) + 0.000 A
score weights: gapWt=0.25, bumpWt=10, HBWt=4
        
subgroup: self dots
atoms selected: 4888
potential dots: 497425
potential area: 31089.1 A^2
  type                 #      %       score score/A^2 x 1000
  C wide_contact     15173   3.1%     156.4      5.03
  C close_contact    14466   2.9%     637.8     20.52
  C small_overlap     4006   0.8%    -122.3     -3.93
  C bad_overlap         47   0.0%      -6.5     -0.21
  C worse_overlap       46   0.0%      -9.1     -0.29
  C H-bonds           1313   0.3%      27.5      0.89
  N wide_contact      4879   1.0%      46.5      1.49
  N close_contact     3237   0.7%     139.8      4.50
  N small_overlap      405   0.1%      -9.9     -0.32
  N bad_overlap          6   0.0%      -0.8     -0.03
  N worse_overlap       12   0.0%      -2.3     -0.07
  N H-bonds           8465   1.7%     237.9      7.65
  O wide_contact     10784   2.2%     112.3      3.61
  O close_contact    11892   2.4%     531.7     17.10
  O small_overlap     3312   0.7%     -95.3     -3.06
  O bad_overlap         34   0.0%      -4.7     -0.15
  O worse_overlap       52   0.0%     -10.1     -0.32
  O H-bonds          35547   7.1%    1049.6     33.76
  P wide_contact       392   0.1%       2.8      0.09
  P close_contact      137   0.0%       5.7      0.18
  P small_overlap       19   0.0%      -0.3     -0.01

     tot contact:    60960  12.3%    1632.9     52.52
     tot overlap:     7939   1.6%    -261.3     -8.40
     tot  H-bond:    45325   9.1%    1315.1     42.30

       grand tot:   114224  23.0%    2686.7     86.42

contact surface area: 7139.0 A^2
```

```
cat 4fenH.compare
B ACT 102 C - radii differ: old is 1.70, new is 1.65
B HPA 101 C2 - radii differ: old is 1.70, new is 1.75
B HPA 101 C4 - radii differ: old is 1.70, new is 1.75
B HPA 101 C5 - radii differ: old is 1.70, new is 1.75
B HPA 101 C6 - radii differ: old is 1.70, new is 1.75
B HPA 101 C8 - radii differ: old is 1.70, new is 1.75
B HPA 101 H2 - radii differ: old is 1.22, new is 1.05
B HPA 101 H8 - radii differ: old is 1.22, new is 1.05
B HPA 101 N1 - flags differ: old is isAcceptor, new is noAcceptor
B HPA 101 N7 - flags differ: old is isAcceptor, new is noAcceptor

```

**Comment by russell-taylor** [comment]
When checking 1xso (which has no ligands so should only be getting info from the standard tables), we get radius 1.70 in the new code rather than 1.65.  This happens on GLN CD and some other CG and CD atoms.  However, these have type 'C" in the model.py type_energy lookup, and those have a radius of 1.65 in the table so it is not clear why these particular ones are 1.70 rather than 1.65 when others are not.  Maybe they don't have the same hydogenation status...

No, hydrogen status is checking for implicit hydrogens which is not set for any of the atoms.   The looked-up value is 1.7.  The radius for C is overridden with 1.65 for many C atoms.  It is not clear why we get 1.70 from a table that should say 1.65.  **It is not overridden because these atom names are CG and CD rather than C.**  We can check this by looking for special amino-acid Carbonyl atoms by adding that function back into AtomTypes.py.  When we do this they match again and 1bti is exactly the same. with 1xso only having different acceptor states for some Nitrogens.

Changing the radius of C from 1.650 to 1.200 in ener_lib.cif does not change the radius used by Probe, and changing O to 1.2 from 1.4 didn't change it either... and removing the ener_lib.cif file completely did not change anything.  Running monomers_to_mon_lib.py after editing that file to change 1.650 to 1.250 did not change anything.  Editing that file to replace a bunch of radii and then rebuilding phenix did not do anything.  It looks like this file is no longer controlling the radii???  Deleting the .pickle and .dlite files and then running mmtbx.rebuild_rotarama_cache and mmtbx.rebuild_cablam_cache also did not change things.  Asked Nigel how to make this file control things as expected.  It turns out that there is another ener_lib.cif file in geostd, which is the one that controls standard residues.

**Comment by russell-taylor** [comment]
(done) Check the differences in 1bti now that we have exactly the same atom information in the old and new cases.  Also check 1xso and others to see where the remaining differences are.

1bti  Most counts are identical or very similar: There are 7 S H-bonds in the new code that are not present in the original.  There are also 57 C H-bonds vs. 50 in the original. The score difference for this is 0.1 and 0.1.  There are 5 more atoms in the new code, which may be Phantom Hydrogens that may be forming these H bonds.

Checking fragments with the waters removed:
- 1bti gets the same total scores and the same individual scores except that the original code pulls out Car atoms separately.
- 1xso has the same number of atoms but a different number of potential dots; it also has four Nitrogens that are acceptors in the new but not in the original code.  Its contact surface area matches to within 2% and its grand total score per unit area is within 0.4 percent.

(done) 1xso shows contacts for the Cu and Zn atoms in the new output, but these are not listed in the original; considered ionically bound in the original?  Jane says to disable all ionic interactions in Probe2 for now by default, adding a Phil parameter to enable them if we want to; they will currently report non-correct interactions (repulsions rather than ionic bonds).  Disabling ionic interactions did not get rid of most of the large set of bumps around ions; it was also noted that the original probe had some wide contacts to one of the Cu atoms so removing the ionic interactions will make the new code behave differently.  Searching for the other cause of this large number of interactions.  They clusters do not appear to directly surround the Cu/Zn atoms (though they are near them).  They do not appear to directly surround the nearby HOH.  A bunch of them are between Nitrogens on neighboring Histidines: NE2 on His 118 A and  ND1 on His 44 A or ND1 on HIS 44 A and NE2 on His 61 A for example (none of these are the ones with different Acceptor status between the old and new, HIS 41 ND1 and NE2).  Opened issue https://github.com/ReliaSolve/cctbx_project/issues/174 for this.

(done) 1xso shows yellow long contact lines labeled HD22 ASN 129 A on one end, and the same name but without a space between the HD22 and ASN on the other end.  There are two of them, and they are in a region of O HOH 261 A vs. HD22 green contact dots.  On the original, the yellow spikes are much shorter.  On the new, they go from near the hemisphere to the separating plane.  The far ends of both old and new appear to be in the same locations visually.  The first coordinate in the output files matches between the old and now; the second point (spike in the code) does not.  The spike is filled in by _save_dot() and is the spikeloc variable in _generate_interaction_dots(); it is in examineDots() in the old code.  The overlap that is used to adjust the radius is computed by DotScorer::check_dot(); that is being used with the proper sign, so it looks like this overlap is what is being computed incorrectly in this case.  The tooCloseHydrogenBond code appears to be adding the correct value to the gap (and subtracting instead makes things worse, as expected); too-close HB seems likely to be the case for these spikes.  Printing the value to use and its addition indicates 0.6 as expected but also no overlaps more than 0.75 or so, so it is not clear how we're getting a spike of length >0.3.  Setting probe.uncharged_hydrogen_cutoff=2.0 made it stop seeing too-close hydrogen bonds at all and made this into a green contact dot, so this is the path that is giving us trouble.  There is some confusion about gap vs. overlap in the return values; we should be using gap (negating it and scaling it) rather than overlap.  Fixing the confusion over overlap vs. gap got the vector to be the right geometry, but it is now hot pink rather than yellow.  Needed to adjust both the gap and the overlap.  It now works.  Added unit test.

(done) The names on the above spike have a space between them in the original Probe line, but not in Probe2.  Fixed by tweaking format commands.

(done) Why are there 5 more Phantom Hydrogens for the new than the old in 1bti?  Is it doing more than one per aromatic ring?  Is it using a larger bonded-check radius?  1xs0 has 35 more out of 5399, 4fenH has 89 more out of 4799.  This is because the original Probe code uses 1.0 (the constant) for determining a couple of lengths when determining which hydrogens to place but the new code used phantomHydrogenRadius (which is set to 1.05) for this.  Switching this to use 1.0 internally makes 1bti counts match between Probe and Probe2.  I've asked Jane what numbers to use for each of these.  Changing them to correct values in old and new has been added as a new issue.

**Comment by russell-taylor** [comment]
(done) Send the kinemage of 1xso for Probe2 so they can look at it

**Comment by russell-taylor** [comment]
**todo:** Add a regression test on whether chem_data changes in a way that changes Probe output (score a molecule fragment).

**Comment by russell-taylor** [comment]
Investigate other differences in dots between Probe and Probe2 running on 1xso_reduced.
- https://github.com/ReliaSolve/cctbx_project/issues/175
- (done) https://github.com/ReliaSolve/cctbx_project/issues/178
- (done) https://github.com/ReliaSolve/cctbx_project/issues/179
- (done) https://github.com/ReliaSolve/cctbx_project/issues/182
- (done) https://github.com/ReliaSolve/cctbx_project/issues/185
- (done) https://github.com/ReliaSolve/cctbx_project/issues/186
- (done) https://github.com/ReliaSolve/cctbx_project/issues/189
- (done) https://github.com/ReliaSolve/cctbx_project/issues/190
- (done) https://github.com/ReliaSolve/cctbx_project/issues/191
- (done) https://github.com/ReliaSolve/cctbx_project/issues/192
- (done) https://github.com/ReliaSolve/cctbx_project/issues/194
- (done) https://github.com/ReliaSolve/cctbx_project/issues/199
- (done) 5fa2 snippet from Jane.
- (done) Re-run visual tests on all fragments to verify that new changes did not break them

**Comment by russell-taylor** [comment]
(done) Why are we getting differences in HIS 41 acceptor states?
```
A HIS 41 ND1 - flags differ: old is noAcceptor, new is isAcceptor
A HIS 41 NE2 - flags differ: old is noAcceptor, new is isAcceptor
B HIS 41 ND1 - flags differ: old is noAcceptor, new is isAcceptor
B HIS 41 NE2 - flags differ: old is noAcceptor, new is isAcceptor
```

We are getting a few extra contacts around at least one water in 1xso, so this seems to be having a (slight) impact.

Modifying the getExtraAtomInfo() code to only mark 'C' atoms in aromatic rings causes a bunch of Nitrogens to be incorrectly marked as not acceptors.

Of all of the Histidines in this file, these are the only ones that have Hydrogens on both of their Nitrogens.  Is there special code in Probe that checks for this and marks them as not acceptors while leaving the Carbons as acceptors?  Perhaps hisNHcheck()?

(done) This will be made moot when https://github.com/ReliaSolve/cctbx_project/issues/176 is solved.  That issue is now in review and we're running our regression tests against it.

**Comment by russell-taylor** [comment]
(done) Add regression test of selection of LEU vs. no selection.
(done) Add regression test of -DROP/-KEEP in both -surf and -self (maybe with resname LEU selection on 3wrp).

**Comment by russell-taylor** [comment]
(done) Test -once and -both on specific portions of specific structure.

**Comment by russell-taylor** [comment]
(done) Why does tst_probe.py work on Windows but fail on Linux?  Verified that we're recompiling on both by adding print statement.  Verified that there are no diffs on either side.

It is failing in Scoring.cpp, reporting "Got no bad bump for bad-bump case" on the third attempt, which should be the first attempt with onlyBumps true (which will have excludeAtoms false).  We get a bump score of 0 in that case in the Linux code.

We're finding the same number of dots and interacting atoms in Windows, but one of them with 2446 dots and 1 interacting atom is not finding any bumps.

Clearing the exclude list explicitly after constructing it does not change the behavior on Linux.  There are no atoms in the excluded list in the one that fails in Linux.

Deleting and rebuilding build/lib and build/iotbx/pdb did not change its behavior.

The dots are all marked as invalid contacts in this case.

(done) They are being marked as ions!  And the element type is U, which is indeed an ion.  We needed to set the element type to a non-ionic one in case we were not setting it rather than assuming it would be filled in with empty.
